### ISSUE TITLE
[Question] Problem with worker access from leader in a cluster

### ISSUE THREAD

**Issue by LWJerri** [caprover/caprover#1628]
Hello everyone! I have two servers that I have clustered together and now I'm having a problem where CapRover can decide to run an application on the second node. When the application tries to connect to the database on the first node, the application gets an error that there is no connection to the database. How do I solve this problem?

The application only throws me an error when it runs on a node that does not have a database specified in the config. I use an internal database connection, i.e. `srv-captain--production-db`. 

Captain dashboard from the "Cluster" section:
![image](https://user-images.githubusercontent.com/50290430/210006126-3d09bf4b-9521-4c72-b643-be915be5c4a7.png)

**Comment by LWJerri** [comment]
All containers with "Exited" status using database connection and can't connect, so CapRover run new container many times until it eventually runs on the node with the database.

![image](https://user-images.githubusercontent.com/50290430/210006476-6b99a078-fab1-4c48-8ca7-cef52b71f5c4.png)

**Comment by githubsaturn** [comment]
> I'm having a problem where CapRover can decide to run an application on the second node

CapRover doesn't decide. It's Docker swarm who makes the decision. 


> When the application tries to connect to the database on the first node, the application gets an error that there is no connection to the database. How do I solve this problem?

Very likely you don't have the correct ports open: 
https://caprover.com/docs/firewall.html

Specifically 7946 and 4789

**Comment by LWJerri** [comment]
> > I'm having a problem where CapRover can decide to run an application on the second node
> 
> CapRover doesn't decide. It's Docker swarm who makes the decision.
> 
> > When the application tries to connect to the database on the first node, the application gets an error that there is no connection to the database. How do I solve this problem?
> 
> Very likely you don't have the correct ports open: https://caprover.com/docs/firewall.html
> 
> Specifically 7946 and 4789

Hello! I run ufw command on all my instances, but nothing changed.

**Comment by LWJerri** [comment]
![image](https://user-images.githubusercontent.com/50290430/210064504-0aeb39e9-1888-404b-98a2-5ba68379a6ee.png)

![image](https://user-images.githubusercontent.com/50290430/210064513-b1e410cd-cf02-43ca-a601-99fab2cad25d.png)

**Comment by LWJerri** [comment]
This is very strange, I've never had any problems with ports, and after I clustered the servers I had them. I used to be able to do port mapping in the app and everything worked without problems, but now I can't do it.

**Comment by githubsaturn** [comment]
1- Just to double check, you have opened the ports on all of your nodes, right? Not just the main node

2- Create a simple new app `test-app`, go to the deployment tab and deploy a simple image like `nginx:alpine` 

3- Keep restarting the app until it ends up being present on the second node. Confirm by running:
```
docker service ps srv-captain--test-app
```
what does it print?

4- shell into captain container and see if you can connect to the test app:
```
docker exec -it $(docker ps --filter name=captain-captain -q) /bin/sh
curl -vI http://test-app
```
what does this print?

**Comment by LWJerri** [comment]
Hello! First of all, Merry Christmas to you. Secondly, I will be able to check all this after the new year, so I would ask you not to close this issue, thank you!

**Comment by LWJerri** [comment]
1.
```text
ubuntu@server-01:~$ sudo ufw status
Status: active

To                         Action      From
--                         ------      ----
80,443,996,2377,3000,4789,7946/tcp ALLOW       Anywhere
2377,4789,7946/udp         ALLOW       Anywhere
8080/tcp                   ALLOW       Anywhere
8080/udp                   ALLOW       Anywhere
80,443,996,2377,3000,4789,7946/tcp (v6) ALLOW       Anywhere (v6)
2377,4789,7946/udp (v6)    ALLOW       Anywhere (v6)
8080/tcp (v6)              ALLOW       Anywhere (v6)
8080/udp (v6)              ALLOW       Anywhere (v6)
```

```text
ubuntu@server-02:~$ sudo ufw status
Status: active

To                         Action      From
--                         ------      ----
80,443,996,2377,3000,4789,7946/tcp ALLOW       Anywhere
2377,4789,7946/udp         ALLOW       Anywhere
8080/tcp                   ALLOW       Anywhere
8080/udp                   ALLOW       Anywhere
80,443,996,2377,3000,4789,7946/tcp (v6) ALLOW       Anywhere (v6)
2377,4789,7946/udp (v6)    ALLOW       Anywhere (v6)
8080/tcp (v6)              ALLOW       Anywhere (v6)
8080/udp (v6)              ALLOW       Anywhere (v6)
```

2. Done.

3.
```text
root@server-01:/home/ubuntu# docker service ps srv-captain--test-app
ID             NAME                          IMAGE                                                       NODE        DESIRED STATE   CURRENT STATE             ERROR     PORTS
hy86f2v5e30w   srv-captain--test-app.1       nginx:alpine                                                server-02   Running         Running 16 seconds ago

4j70ox58vl50    \_ srv-captain--test-app.1   registry.01.lwjerri.ml:996/captain/img-captain-test-app:0   server-01   Shutdown        Shutdown 13 seconds ago

0x68cxpwwb0q    \_ srv-captain--test-app.1   caprover/caprover-placeholder-app:latest                    server-02   Shutdown        Shutdown 56 seconds ago
```

4.
```text
root@server-01:/home/ubuntu# docker exec -it $(docker ps --filter name=captain-captain -q) /bin/sh
# curl -vI http://test-app
* Expire in 0 ms for 6 (transfer 0xaaaaf5474060)
* Expire in 1 ms for 1 (transfer 0xaaaaf5474060)
* Expire in 0 ms for 1 (transfer 0xaaaaf5474060)
* Expire in 1 ms for 1 (transfer 0xaaaaf5474060)
* Expire in 0 ms for 1 (transfer 0xaaaaf5474060)
```

When app running on `server-02`, I get blue screen with 502 error, but on main `server-01` all works good.

**Comment by LWJerri** [comment]
I have Adminer running, when it runs on the second server I can't use it, I just see a blue screen with a 502 error, but when the container runs on the first server everything is fine.

**Comment by LWJerri** [comment]
I also use my servers on Oracle Cloud Instance, but I have all ports allowed there. @githubsaturn 

![image](https://user-images.githubusercontent.com/50290430/210222265-dcf73093-37a3-4172-938e-15478c1de244.png)
![image](https://user-images.githubusercontent.com/50290430/210222275-9141a191-49f5-40bc-a70c-1a85f0da0565.png)

**Comment by LWJerri** [comment]
I just checked the port status for my two servers and it turns out that all servers have port 4789 closed...

**Comment by githubsaturn** [comment]
As mentioned here: https://github.com/caprover/caprover/issues/1628#issuecomment-1367716203

These ports are used by Docker to connect containers on different nodes. 

> I just checked the port status for my two servers and it turns out that all servers have port 4789 closed...

This is exactly the root cause. You need to make sure that these ports are open.

**Comment by LWJerri** [comment]
I mean, nobody uses port 4789.

**Comment by LWJerri** [comment]
> As mentioned here: [#1628 (comment)](https://github.com/caprover/caprover/issues/1628#issuecomment-1367716203)
> 
> These ports are used by Docker to connect containers on different nodes.
> 
> > I just checked the port status for my two servers and it turns out that all servers have port 4789 closed...
> 
> This is exactly the root cause. You need to make sure that these ports are open. Glad to hear that the issue is resolved.

Here (https://github.com/caprover/caprover/issues/1628#issuecomment-1368836599) you can see, I have already openned port.

**Comment by githubsaturn** [comment]
I don't understand. Opening a port has nothing to do with "using it" Docker swarm uses the udp ports. 

> but if I open the port in some application

What does this mean? You can't "open" a port in any application? Are you referring to mapping a port?

**Comment by githubsaturn** [comment]
Also, where does it show open/closed status that you're mentioning here:?
> so they show up as closed, but if I open the port in some application, it shows me that the port is open.

**Comment by LWJerri** [comment]
When checking any port, it will appear as "closed" until I run something on that port. The same goes for port 4789. It wasn't used by anyone, so it showed up as closed, but when I did a port mapping in the application with database 4789 -> 5432, it started showing up as "open".

**Comment by LWJerri** [comment]
> Also, where does it show open/closed status that you're mentioning here:?
> 
> > so they show up as closed, but if I open the port in some application, it shows me that the port is open.

https://www.yougetsignal.com/tools/open-ports/

**Comment by githubsaturn** [comment]
That tool is not an accurate way of understanding open/close status. 

What's the ip address of your servers?

**Comment by LWJerri** [comment]
> That tool is not an accurate way of understanding open/close status.
> 
> What's the ip address of your servers?

server-02: 130.61.249.122
server-01: 130.61.223.243

**Comment by githubsaturn** [comment]
Okay, let me compile a few tests and send it your way.

**Comment by LWJerri** [comment]
> Okay, let me compile a few tests and send it your way.

Okay, np.

**Comment by githubsaturn** [comment]
Just realized that you're using arm CPU... This might be a bug with docker engine. Let's try it out:

1) ensure you have `dig` installed on your `captain-captain`
```
docker exec -it $(docker ps --filter name=captain-captain -q) /bin/bash -c "apt update && apt install dnsutils -yqq"
```

2) ensure you can view `captain-nginx` from `captain-captain`
```
docker exec -it $(docker ps --filter name=captain-captain -q) /bin/bash -c "dig +short captain-nginx"
```

3) 
[this is the part where you could see the problem]
Pick a service that is on server-02, then run a dig:

```
docker exec -it $(docker ps --filter name=captain-captain -q) /bin/bash -c "dig +short srv-captain--your-app-name-on-server02"
```


If this last command prints an error, then it's a Docker issue. Let me know if that's the case.

**Comment by LWJerri** [comment]
> Just realized that you're using arm CPU... This might be a bug with docker engine. Let's try it out:
> 
> 1. ensure you have `dig` installed on your `captain-captain`
> 
> ```
> docker exec -it $(docker ps --filter name=captain-captain -q) /bin/bash -c "apt update && apt install dnsutils -yqq"
> ```
> 
> 2. ensure you can view `captain-nginx` from `captain-captain`
> 
> ```
> docker exec -it $(docker ps --filter name=captain-captain -q) /bin/bash -c "dig captain-nginx"
> ```
> 
> 3. Check an app on sever-02. This is the part where you could see the problem.
>    Pick a service that is on server-02, then run a dig:
> 
> ```
> docker exec -it $(docker ps --filter name=captain-captain -q) /bin/bash -c "dig srv-captain--your-app-name-on-server02"
> ```
> 
> If this last command prints an error, then it's a Docker issue. Let me know if that's the case.

```text
root@server-01:/home/ubuntu# docker exec -it $(docker ps --filter name=captain-captain -q) /bin/bash -c "dig captain-nginx"

; <<>> DiG 9.11.5-P4-5.1+deb10u8-Debian <<>> captain-nginx
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 18241
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;captain-nginx.                 IN      A

;; ANSWER SECTION:
captain-nginx.          600     IN      A       10.0.1.6

;; Query time: 0 msec
;; SERVER: 127.0.0.11#53(127.0.0.11)
;; WHEN: Mon Jan 02 18:21:06 UTC 2023
;; MSG SIZE  rcvd: 60

root@server-01:/home/ubuntu# docker exec -it $(docker ps --filter name=captain-captain -q) /bin/bash -c "dig srv-captain--test-app"

; <<>> DiG 9.11.5-P4-5.1+deb10u8-Debian <<>> srv-captain--test-app
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 33264
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;srv-captain--test-app.         IN      A

;; AUTHORITY SECTION:
.                       3592    IN      SOA     a.root-servers.net. nstld.verisign-grs.com. 2023010200 1800 900 604800 86400

;; Query time: 0 msec
;; SERVER: 127.0.0.11#53(127.0.0.11)
;; WHEN: Mon Jan 02 18:23:25 UTC 2023
;; MSG SIZE  rcvd: 125
```

**Comment by LWJerri** [comment]
```text
root@server-02:/home/ubuntu# docker ps
CONTAINER ID   IMAGE          COMMAND                  CREATED         STATUS         PORTS     NAMES
08f5d51559b1   nginx:alpine   "/docker-entrypoint.…"   2 minutes ago   Up 2 minutes   80/tcp    srv-captain--test-app.1.zp92dvigyr734ly8e4yueqpk6
```

```text
root@server-01:/home/ubuntu# docker exec -it $(docker ps --filter name=captain-captain -q) /bin/bash -c "dig srv-captain--test-app.1.zp92dvigyr734ly8e4yueqpk6"

; <<>> DiG 9.11.5-P4-5.1+deb10u8-Debian <<>> srv-captain--test-app.1.zp92dvigyr734ly8e4yueqpk6
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 26219
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1372
;; QUESTION SECTION:
;srv-captain--test-app.1.zp92dvigyr734ly8e4yueqpk6. IN A

;; AUTHORITY SECTION:
.                       3598    IN      SOA     a.root-servers.net. nstld.verisign-grs.com. 2023010200 1800 900 604800 86400

;; Query time: 2 msec
;; SERVER: 127.0.0.11#53(127.0.0.11)
;; WHEN: Mon Jan 02 18:26:25 UTC 2023
;; MSG SIZE  rcvd: 153
```

**Comment by githubsaturn** [comment]
Yup - the second command failed. Like what I thought, it's a Docker issue. Feel free to follow up with them on what is causing this.

![image](https://user-images.githubusercontent.com/10582082/210267749-1de9444e-2b9c-4eb1-b5d1-dac5963d6636.png)



File an issue on [github.com/moby/moby](https://github.com/moby/moby/issues)


Copy paste the output these commands for the Docker folks to be able to debug your issue:
```
docker service inspect captain-captain
docker service inspect srv-captain--test-app
docker network inspect captain-overlay-network
docker service ps srv-captain--test-app
```

**Comment by LWJerri** [comment]
> Yup - the second command failed. Like what I thought, it's a Docker issue. Feel free to follow up with them on what is causing this.
> 
> ![image](https://user-images.githubusercontent.com/10582082/210267749-1de9444e-2b9c-4eb1-b5d1-dac5963d6636.png)
> 
> File an issue on [github.com/moby/moby](https://github.com/moby/moby/issues)
> 
> Copy paste the output these commands for the Docker folks to be able to debug your issue:
> 
> ```
> docker service inspect captain-captain
> docker service inspect srv-captain--test-app
> docker network inspect captain-overlay-network
> docker service ps srv-captain--test-app
> ```

Should I just create an ishua there, link to your comment and prefix the output of the commands?

**Comment by githubsaturn** [comment]
Yes, create an issue there, link this issue, but it's too long for them to read, so just to help them read, I suggest summerize the issue:

Just tell them:
Cannot connect to the app if it's on a different node. Proof:
```
docker exec -it $(docker ps --filter name=captain-captain -q) /bin/bash -c "dig srv-captain--your-app-name-on-server02"
```

Also more details:
copy paste the outputs from these commands:
```
docker service inspect captain-captain
docker service inspect srv-captain--test-app
docker network inspect captain-overlay-network
docker service ps srv-captain--test-app
```

**Comment by githubsaturn** [comment]
To be clear, this is not a CapRover issue at this point, since even without CapRover the two containers are failing to connect.

**Comment by LWJerri** [comment]
Done.
